src/libproc/proc_pid.rs

```diff
@@ -503,3 +503,132 @@ fn name_test_init_pid() {
         Err(message) => assert!(true, message)
     }
 }
+
+// This trait is needed for polymorphism on listpidinfo types, also abstracting flavor in order to provide
+// type-guaranteed flavor correctness
+pub trait ListPIDInfo {
+    type Item;
+    fn flavor() -> PidInfoFlavor;
+}
+
+/// Returns the information of the process that match pid passed in.
+/// `max_len` is the maximum number of array to return.
+/// The length of return value: `Vec<T::Item>` may be less than `max_len`.
+///
+/// # Examples
+///
+/// ```
+/// use std::io::Write;
+/// use libproc::libproc::proc_pid::{listpidinfo, pidinfo, ListFDs, ListThreads, TaskAllInfo};
+///
+/// fn listthreads_test() {
+///     use std::process;
+///     let pid = process::id() as i32;
+///
+///     match pidinfo::<TaskAllInfo>(pid, 0) {
+///         Ok(info) => {
+///             match listpidinfo::<ListThreads>(pid, info.ptinfo.pti_threadnum as usize) {
+///                 Ok(threads) => assert!(threads.len()>0),
+///                 Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///             }
+///             match listpidinfo::<ListFDs>(pid, info.pbsd.pbi_nfiles as usize) {
+///                 Ok(fds) => assert!(fds.len()>0),
+///                 Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///             }
+///         },
+///         Err(err) => assert!(false, "Error retrieving process info: {}", err)
+///     };
+/// }
+/// ```
+pub fn listpidinfo<T: ListPIDInfo>(pid : i32, max_len: usize) -> Result<Vec<T::Item>, String> {
+    let flavor = T::flavor() as i32;
+    let buffer_size = mem::size_of::<T::Item>() as i32 * max_len as i32;
+    let mut buffer = Vec::<T::Item>::with_capacity(max_len);
+    let buffer_ptr = unsafe {
+        buffer.set_len(max_len);
+        buffer.as_mut_ptr() as *mut c_void
+    };
+
+    let ret: i32;
+
+    unsafe {
+        ret = proc_pidinfo( pid, flavor, 0, buffer_ptr, buffer_size);
+    };
+
+    if ret <= 0 {
+        Err(get_errno_with_message(ret))
+    } else {
+        let actual_len = ret as usize / mem::size_of::<T::Item>();
+        buffer.truncate(actual_len);
+        Ok(buffer)
+    }
+}
+
+#[test]
```

Comment:
Do we need the unit test if you have the same code in the doc comment above?
It will get tested when the doc tests get run.

Reply:
Surely test is duplicated. I'll remove it.

Reply:
I'd either:
- remove the duplication
- leave the unit tests with asserts!() and change the doc comment to be more of an example of how to use it

Reply:
I reverted unit test and I changed doc comment.